File: ios/native/MauiEmarsys/Mapper/MessageMapper.swift
//
//  Copyright © 2026 Emarsys. All rights reserved.
//

import Foundation
import EmarsysSDK

class MessageMapper {
    static func mapMessages(_ messages: [EMSMessage]) -> [[String: Any]] {
        let messageDicts: [[String: Any]] = messages.map { message in
            var messageDict = [String: Any]()
            messageDict["id"] = message.id
            messageDict["campaignId"] = message.campaignId
            messageDict["collapseId"] = message.collapseId
            messageDict["title"] = message.title
            messageDict["body"] = message.body
            messageDict["imageUrl"] = message.imageUrl
            messageDict["receivedAt"] = message.receivedAt
            messageDict["updatedAt"] = message.updatedAt
            messageDict["expiresAt"] = message.expiresAt
            messageDict["tags"] = message.tags
            messageDict["properties"] = message.properties
            if let actions = message.actions {
                messageDict["actions"] = mapActions(actions)
            }
            return messageDict
        }
        return messageDicts
    }
    
    static func mapActions(_ actions: [EMSActionModelProtocol]) -> [[String: Any]] {
        let actionDicts: [[String: Any]] = actions.map { action in
            var actionDict = [String: Any]()
            if let appEvent = action as? EMSAppEventActionModel {
                actionDict["id"] = appEvent.id
                actionDict["title"] = appEvent.title
                actionDict["type"] = appEvent.type
                actionDict["name"] = appEvent.name
                actionDict["payload"] = appEvent.payload
            } else if let customEvent = action as? EMSCustomEventActionModel {
                actionDict["id"] = customEvent.id
                actionDict["title"] = customEvent.title
                actionDict["type"] = customEvent.type
                actionDict["name"] = customEvent.name
                actionDict["payload"] = customEvent.payload
            } else if let externalUrl = action as? EMSOpenExternalUrlActionModel {
                actionDict["id"] = externalUrl.id
                actionDict["title"] = externalUrl.title
                actionDict["type"] = externalUrl.type
                actionDict["url"] = externalUrl.url.absoluteString
            }
            return actionDict
        }
        return actionDicts
    }
}

File: android/native/emarsys/src/main/java/com/emarsys/maui/mapper/MessageMapper.java
package com.emarsys.maui.mapper;

import androidx.annotation.NonNull;
import com.emarsys.mobileengage.api.action.ActionModel;
import com.emarsys.mobileengage.api.action.AppEventActionModel;
import com.emarsys.mobileengage.api.action.CustomEventActionModel;
import com.emarsys.mobileengage.api.action.OpenExternalUrlActionModel;
import com.emarsys.mobileengage.api.inbox.Message;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class MessageMapper {

    public static @NonNull List<Map<String, Object>> mapMessages(@NonNull List<Message> messages) {
        return messages.stream().map(message -> {
            Map<String, Object> resultMap = new HashMap<>();
            resultMap.put("id", message.getId());
            resultMap.put("campaignId", message.getCampaignId());
            if (message.getCollapseId() != null) {
                resultMap.put("collapseId", message.getCollapseId());
            }
            resultMap.put("title", message.getTitle());
            resultMap.put("body", message.getBody());
            if (message.getImageUrl() != null) {
                resultMap.put("imageUrl", message.getImageUrl());
            }
            resultMap.put("receivedAt", message.getReceivedAt());
            if (message.getUpdatedAt() != null) {
                resultMap.put("updatedAt", message.getUpdatedAt());
            }
            if (message.getExpiresAt() != null) {
                resultMap.put("expiresAt", message.getExpiresAt());
            }
            if (message.getTags() != null) {
                resultMap.put("tags", message.getTags());
            }
            if (message.getProperties() != null) {
                resultMap.put("properties", message.getProperties());
            }
            if (message.getActions() != null) {
                resultMap.put("actions", mapActions(message.getActions()));
            }
            return resultMap;
        }).collect(Collectors.toList());
    }

    public static @NonNull List<Map<String, Object>> mapActions(@NonNull List<ActionModel> actions) {
        return actions.stream().map(action -> {
            Map<String, Object> actionMap = new HashMap<>();
            actionMap.put("id", action.getId());
            actionMap.put("title", action.getTitle());
            actionMap.put("type", action.getType());
            if (action instanceof AppEventActionModel) {
                AppEventActionModel appEvent = (AppEventActionModel) action;
                actionMap.put("name", appEvent.getName());
                actionMap.put("payload", appEvent.getPayload());
            } else if (action instanceof CustomEventActionModel) {
                CustomEventActionModel customEvent = (CustomEventActionModel) action;
                actionMap.put("name", customEvent.getName());
                actionMap.put("payload", customEvent.getPayload());
            } else if (action instanceof OpenExternalUrlActionModel) {
                OpenExternalUrlActionModel externalUrl = (OpenExternalUrlActionModel) action;
                actionMap.put("url", externalUrl.getUrl().toString());
            }
            return actionMap;
        }).collect(Collectors.toList());
    }

}
